### PR TITLE
[api-replacement] replace json_rpc_invocations with Dev API in deploy.rs

### PR DESCRIPTION
**Pull request by avin1100**
## Motivation

Currently, we are using a hard coded json-rpc value to deploy modules onchain. We want to remove all hard-coded json-rpc values from shuffle and replace them with the dev API. This PR replaces the json rpc value in deploy.rs with the dev API. 

## Test Plan

The goal is to mimic what deploy currently does. So I ran shuffle console and inside the console ran Shuffle.accountTransactions(). Notice how there are no transactions since I haven't deployed anything yet. 

Screen Shot 2021-10-28 at 10 48 52 AM

Then I ran shuffle deploy "project_path". Notice that 3 modules have deployed and hence 3 transactions have been created.  

image

Now when I rerun Shuffle.accountTransactions() in console, I get these transactions. Notice the hash on the first transaction matches the hash of the first deployed module transaction in the picture above. 

Screen Shot 2021-10-28 at 10 50 59 AM

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: shuffle/cli/src/deploy.rs [hunk 1]
```diff
@@ -1,50 +1,40 @@
...
+use std::{
+    path::Path,
+    time::{Duration, Instant},
+};
 
 /// Deploys shuffle's main Move Package to the sender's address.
-pub fn handle(project_path: &Path) -> Result<()> {
-    let home = Home::new(shared::get_home_path().as_path())?;
+pub async fn handle(project_path: &Path) -> Result<()> {
+    let network = normalized_network(read_config(project_path)?.get_network())?;
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:28]: This will be outdated with your networks.toml PR.

File: shuffle/cli/src/deploy.rs [hunk 2]
```diff
@@ -53,65 +43,223 @@ fn publish_packages_as_transaction(
...
+) -> Result<()> {
+    for hash in all_hashes.iter() {
+        let mut resp = client.get_transactions_by_hash(hash.as_str()).await?;
+        let mut json: serde_json::Value = serde_json::from_str(resp.text().await?.as_str())?;
+        let start = Instant::now();
+        while json["type"] == "pending_transaction" {
+            resp = client.get_transactions_by_hash(hash.as_str()).await?;
+            json = serde_json::from_str(resp.text().await?.as_str())?;
+            let duration = start.elapsed();
+            if duration > Duration::from_secs(10) {
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:106]: What's going on here? Are you spinning in the while loop until 10 seconds pass or json type is != `pending_transaction`? That's a lot of CPU spins, let's sleep a bit.

File: shuffle/cli/src/deploy.rs [hunk 3]
```diff
@@ -53,65 +43,223 @@ fn publish_packages_as_transaction(
...
+    Ok(())
+}
+
+fn is_execution_successful(json: &Value) -> Result<bool> {
+    json["success"]
+        .as_bool()
+        .ok_or_else(|| anyhow!("Unable to access success key"))
+}
+
+pub async fn confirm_modules_exist(
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:130]: Why do we need this at all? If `is_execution_successful` is true, we can move on. Let's ✂️ this whole method.

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/deploy.rs (hunk 1 above)
Comment (avin1100) [reply shuffle/cli/src/deploy.rs:28]: Agreed. Will refactor this code after networks.toml PR lands

File: shuffle/cli/src/deploy.rs (hunk 2 above)
Comment (avin1100) [reply shuffle/cli/src/deploy.rs:106]: I'm spinning in the while loop until  json type is != `pending_transaction`. If for whatever reason, json type stays at pending_transactions, after 10 seconds we break out of the loop. Will add a sleep here too.

File: shuffle/cli/src/deploy.rs (hunk 3 above)
Comment (avin1100) [reply shuffle/cli/src/deploy.rs:130]: Confirm module exists? I thought so too, but I got this comment from my last (outdated) replace json_rpc with DEV API PR: https://github.com/diem/diem/pull/9565/files#r738786745 . From here I worked on making another check to see if the module exists.

Comment (avin1100) [reply shuffle/cli/src/deploy.rs:130]: Spoke to Xiao and he informed me that If txn is deploy a module and txn executed successfully, then if module does not exist, it means there is a bug. So i'll remove this extra confirm_module_exist check

File: shuffle/cli/src/deploy.rs [hunk 4]
```diff
@@ -53,65 +38,165 @@ fn publish_packages_as_transaction(
...
+    );
+    let bytes = bcs::to_bytes(&publish_txn)?;
+    let resp = client.post_transactions(bytes).await?;
+    let json: serde_json::Value = serde_json::from_str(resp.text().await?.as_str())?;
+    let hash = get_hash_from_post_txn(json)?;
+    Ok(hash)
+}
 
-            shared::send_transaction(client, publish_txn)?;
+pub async fn confirm_txns_executed_from_hashes(
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:84]: 1. Why is this public? Let's remove the `pub`
2. This function is still too dense, let's break it out into more testable methods. Suggestions: pull out what's inside the `for hash in all_hashes`
3. `time::Duration::from_millis(1000)` => `time::Duration::from_second(1)`
There's also a world where you check the hash execution in parallel and join at the end.

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/deploy.rs [hunk 5]
```diff
@@ -53,65 +38,165 @@ fn publish_packages_as_transaction(
...
+            thread::sleep(time::Duration::from_millis(1000));
+            resp = client.get_transactions_by_hash(hash.as_str()).await?;
+            json = serde_json::from_str(resp.text().await?.as_str())?;
+            let duration = start.elapsed();
+            if duration > Duration::from_secs(10) {
+                break;
+            }
+        }
+        if is_execution_successful(&json)? {
+            println!("Transaction with hash {} executed successfully", hash);
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:102]: Let's only print on error, it's too noisy to print every success.

image

File: shuffle/cli/src/deploy.rs [hunk 6]
```diff
@@ -53,65 +38,165 @@ fn publish_packages_as_transaction(
...
         } else {
-            println!("Skipping Module: {}", module_id);
+            println!(
+                "Transaction with hash {} didn't execute successfully:",
+                hash
+            );
```
Comment (dimroc) [inline shuffle/cli/src/deploy.rs:107]: If there is a failure, we should return `Err(...)`. Ideally, we would aggregate all errors, but it's fine to error on the first. Let's bubble up any useful information we have. ideas are vm_status, error message, wtever.

**Review by dimroc — CHANGES_REQUESTED** [review]

**Review by dimroc — DISMISSED** [review]

**Pull request closed without merging**